File: app/src/main/java/com/touchnothing/app/domain/UserStatsService.kt
package com.touchnothing.app.domain

import com.touchnothing.app.data.model.UserStats
import com.touchnothing.app.data.remote.SupabaseService

class UserStatsService(
    private val supabaseService: SupabaseService = SupabaseService.shared,
) {
    suspend fun load(nickname: String?): UserStatsLoadResult {
        if (nickname == null) {
            return UserStatsLoadResult(stats = UserStats.Empty, status = UserStatsLoadStatus.MISSING_NICKNAME)
        }
        return runCatching {
            val stats = supabaseService.fetchUserStats(nickname)
            UserStatsLoadResult(stats = stats, status = UserStatsLoadStatus.SUCCESS)
        }.getOrElse {
            UserStatsLoadResult(stats = UserStats.Empty, status = UserStatsLoadStatus.FAILED)
        }
    }
}

File: TouchNothing/Services/UserStatsService.swift
import Foundation

enum UserStatsLoadStatus: Equatable {
    case success
    case failed
    case missingNickname
}

struct UserStatsLoadResult {
    let stats: UserStats
    let status: UserStatsLoadStatus
}

enum UserStatsService {
    static func load(nickname: String?) async -> UserStatsLoadResult {
        guard let nickname else {
            return UserStatsLoadResult(stats: .empty, status: .missingNickname)
        }

        do {
            let stats = try await SupabaseService.shared.fetchUserStats(nickname: nickname)
            return UserStatsLoadResult(stats: stats, status: .success)
        } catch {
            return UserStatsLoadResult(stats: .empty, status: .failed)
        }
    }
}
